editor/plugins/material_editor_plugin.cpp

```diff
@@ -35,44 +35,87 @@
-	if (p_what == NOTIFICATION_READY) {
-		//get_scene()->connect("node_removed",this,"_node_removed");
-
-		if (first_enter) {
-			//it's in propertyeditor so.. could be moved around
-
-			light_1_switch->set_normal_texture(get_theme_icon("MaterialPreviewLight1", "EditorIcons"));
-			light_1_switch->set_pressed_texture(get_theme_icon("MaterialPreviewLight1Off", "EditorIcons"));
-			light_2_switch->set_normal_texture(get_theme_icon("MaterialPreviewLight2", "EditorIcons"));
-			light_2_switch->set_pressed_texture(get_theme_icon("MaterialPreviewLight2Off", "EditorIcons"));
-
-			sphere_switch->set_normal_texture(get_theme_icon("MaterialPreviewSphereOff", "EditorIcons"));
-			sphere_switch->set_pressed_texture(get_theme_icon("MaterialPreviewSphere", "EditorIcons"));
-			box_switch->set_normal_texture(get_theme_icon("MaterialPreviewCubeOff", "EditorIcons"));
-			box_switch->set_pressed_texture(get_theme_icon("MaterialPreviewCube", "EditorIcons"));
-
-			first_enter = false;
+void MaterialEditor::_gui_input(Ref<InputEvent> p_event) {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventMouseMotion> mm = p_event;
+	if (mm.is_valid() && mm->get_button_mask() & MOUSE_BUTTON_MASK_LEFT) {
+		rot_x -= mm->get_relative().y * 0.01;
+		rot_y -= mm->get_relative().x * 0.01;
+		if (rot_x < -Math_PI / 2) {
+			rot_x = -Math_PI / 2;
+		} else if (rot_x > Math_PI / 2) {
+			rot_x = Math_PI / 2;
 		}
+		_update_rotation();
 	}
+}
 
-	if (p_what == NOTIFICATION_DRAW) {
-		Ref<Texture2D> checkerboard = get_theme_icon("Checkerboard", "EditorIcons");
-		Size2 size = get_size();
-
-		draw_texture_rect(checkerboard, Rect2(Point2(), size), true);
+void MaterialEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			if (first_enter) {
+				light_1_switch->set_normal_texture(get_theme_icon("MaterialPreviewLight1", "EditorIcons"));
+				light_1_switch->set_pressed_texture(get_theme_icon("MaterialPreviewLight1Off", "EditorIcons"));
+				light_2_switch->set_normal_texture(get_theme_icon("MaterialPreviewLight2", "EditorIcons"));
+				light_2_switch->set_pressed_texture(get_theme_icon("MaterialPreviewLight2Off", "EditorIcons"));
+
+				sphere_switch->set_normal_texture(get_theme_icon("MaterialPreviewSphereOff", "EditorIcons"));
+				sphere_switch->set_pressed_texture(get_theme_icon("MaterialPreviewSphere", "EditorIcons"));
+				box_switch->set_normal_texture(get_theme_icon("MaterialPreviewCubeOff", "EditorIcons"));
+				box_switch->set_pressed_texture(get_theme_icon("MaterialPreviewCube", "EditorIcons"));
+				first_enter = false;
+			}
+		} break;
+		case NOTIFICATION_DRAW: {
+			Ref<Texture2D> checkerboard = get_theme_icon("Checkerboard", "EditorIcons");
+			Size2 size = get_size();
+
+			draw_texture_rect(checkerboard, Rect2(Point2(), size), true);
+		} break;
 	}
 }
 
+void MaterialEditor::_update_rotation() {
+	Transform3D t;
+	t.basis.rotate(Vector3(0, 1, 0), -rot_y);
+	t.basis.rotate(Vector3(1, 0, 0), -rot_x);
```

Comment:
I am not good at basis transformations.

I tried the way you did.
t.basis = Basis(Vector3(sun_rotation.x, sun_rotation.y, 0));

like this,
t.basis = Basis(Vector3(rot.x, rot.y, 0));

And all other combinations(changing signs).
None of them worked in the expected manner.

So, I left like what it was.

Reply:
> Just do it [like this](https://github.com/godotengine/godot/pull/46462/files#diff-57832e09092105245d5eaa89d7834a30f5c6bd6e5b5166450fde12f61a585d4cR6560-R6563).

I couldn't figure it out yet. So, I am leaving it as it. Since, mesh editors preview code is same as here, It shouldn't make any difference. If you wanna do your way,then you can continue this PR :D